hibernate-core/src/main/java/org/hibernate/dialect/H2Dialect.java

```diff
@@ -453,10 +453,8 @@ public String timestampaddPattern(TemporalUnit unit, TemporalType temporalType,
 			return "(?2+?3)";
 		}
 		return unit == SECOND
-				//TODO: if we have an integral number of seconds
-				//      (the common case) this is unnecessary
-				? "dateadd(nanosecond,?2*1e9,cast(?3 as timestamp))"
-				: "dateadd(?1,?2,cast(?3 as timestamp))";
+				? "dateadd(nanosecond,?2*1e9,cast(?3 as " + temporalType.name().toLowerCase() + "))"
+				: "dateadd(?1,?2,?3)";
```

Comment:
@cigaly we need to be a bit careful here, since this has diverged on the H7 branch. Over there, I changed it to:

```java
				? "dateadd(nanosecond,?2*1e9,?3)"
				: "dateadd(?1,?2,?3)";
```

It looks like I decided we didn't need the `cast()` at all.

Does that work?

Reply:
Unfortunately no, when ?3 is parameter it is creating query like

```
select count(*) 
from SimpleEntity se1_0 
where se1_0.inst>dateadd(nanosecond,2*1e9,dateadd(nanosecond,1*1e9,?))
```

and then throwing

`org.hibernate.exception.GenericJDBCException: could not prepare statement [Unknown data type: "?"; `

Additional casting is only needed in situation when temporal is parameter. Otherwise, existing implementations are good.

Reply:
OK, fair enough